### PR TITLE
[Refine] New robust downloading mechanism for otaclient

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Introduction
This PR introduces a much robust **time-based** downloading mechanism for otaclient downloading anything from remote servers. 
Comparing to the previous implementation that fails on single failed download task, the new strategy now **changes to keep retrying forever, unless the downloading keeps failing(no new succeeded task) longer than the configured timeout.** 

**The behavior is as follow:**
If the downloading failures occurs but there are still tasks succeed, _like the network environment is unstable but still connected,_ the otaclient will keep retrying until all the downloading tasks succeed.
But if downloading keeps failing, _like the user disconnects the Internet to the device_, otaclient will keep retrying **until timeout reached**.

Comparing to the previous implementation that fails on single download task, the new downloading mechanism introduces the concept of **downloading tasks group**, manage all downloading tasks together, and retry based on **tasks group** scope.  New downloading retry strategy is implemented with the use of **downloading tasks group**. 

## Major changes
1. new `RetryTaskMap` class: works like built-in `map`, take a `func` and an `iterable` as input, and it will try its best to finish all the tasks. This class is used in downloading remote OTA files and downloading OTA image meta files to implement the new robust downloading mechanism. Also the `create_standby.rebuild_mode` module uses this class too.
2. `create_standby` module interface is updated, now it will no-longer be in charge of downloading anything from remote, but only focus on calculating **delta** and apply update to standby slot.
3. otaclient now pre-calculates and prepare the delta before downloading and applying update. With this change, otaclient can report the amount and the size of files needed to be downloaded before downloading starts. 
4. `OTAClient` now takes charge of processing `metadata.jwt`, downloading OTA image meta files and downloading needed OTA image files. A new `DOWNLOADING` phase is implemented and the OTA image files downloading will be done during this phase, this phase occurs before update to standby slot, after the meta files processing and delta calculation.
5. new downloading retry strategy is implemented with using `RetryTaskMap`. The retry strategy now changes to keep retrying **forever**, unless the downloading keep failing(no new succeeded task) longer than the **DOWNLOAD_GROUP_NO_SUCCESS_RETRY_TIMEOUT**. This allow the user to manually interrupt the downloading by disconnect the network.
6. create_standby: deprecate `LEGACY` mode. Now only `REBUILD` mode is available.

## Other changes
1. configs: refine and re-organize, split the internal used config fields(which users should not change) into `_InternalSettings` class, and users should only change the fields in `BasiceConfig` class.
2. configs: add new config fields for new downloading mechanism, especially the `DOWNLOAD_GROUP_NO_SUCCESS_RETRY_TIMEOUT = 5 * 60`, the otaclient will give up retrying if keep failing longer than 5 mins without any new task succeeded.
3. configs: reduce single download task retry to 3 as we now mainly relies on the downloading tasks group scope retry.
4. common: deprecate `SimpleTaskTracker` class, used `RetryTaskMap` instead.
5. create_standby: fully implement `DeltaGenerator`, resolve the leftover work. Now it will create a full `DeltaBundle` after delta calculation is finished.
6. create_standby: update interface definition, add a new `calculate_and_prepare_delta` API method. Deprecate `is_standby_as_ref` method.
7. ota_client: no-longer reused `_OTAUpdater` instance, instead always using new inst for handling new update request.
8. update_stats: define new `OP_DOWNLOAD_FAILED_REPORT` for reporting failed single downloading task.
9. re-implement all related test files.

## Bug fixed
1. ota_client: fix update_lock being released unexpectedly when multiple update requests received

## Potential future works
1. report the download size before starting download.
2. . refinement over the definition of `StatusProgressPhase`.
3. . in-place update mode now can be implemented as the `DeltaGenerator` is fully implemented.

## Tests

- [x] test files
- [x] test on VM(with otaproxy enabled)
- [x] test on VM(with otaproxy enabled, 5mins network disconnect test)
- [x] test on VM(with otaproxy enabled, long time disconnect and resume test) 

## Ticket
[T4PB-24839 Introducing a new robust retry strategy](https://tier4.atlassian.net/browse/T4PB-24839)
(partially related) [T4PB-24097 multiple ECUs OTA update failure handling](https://tier4.atlassian.net/browse/T4PB-24097)

### CODE REVIEW

**Comment by Bodong-Yang** [comment]
@obi-t4 This PR is again relatively large(sorry for that), so I recommend to first run the test files and test with VM on your side, and then start the review along with reading the PR introduction. Thank you!

File: otaclient/app/configs.py [hunk 1]
```diff
@@ -77,44 +64,86 @@ class BaseConfig:
     PASSWD_FILE = "/etc/passwd"
     GROUP_FILE = "/etc/group"
     FSTAB_FPATH = "/etc/fstab"
+    # where the OTA image meta store for this slot
+    META_FOLDER = "/opt/ota/image-meta"
```
Comment (obi-t4) [inline otaclient/app/configs.py:68]: folder and directory are the same meaning and I'd like to unify this terminology to `directory`, in the future. Make sure not this PR is OK.

File: otaclient/app/create_standby/rebuild_mode.py [hunk 2]
```diff
@@ -61,95 +42,52 @@
...
+        delta_calculator = DeltaGenerator(
+            delta_src=self.active_slot_mp,
+            local_copy_dir=self._ota_tmp,
+            stats_collector=self.stats_collector,
+        )
+        delta_bundle = delta_calculator.calculate_and_process_delta(
+            delta_src_reg=Path(cfg.META_FOLDER) / regular_inf_fname,
+            new_reg=self._ota_tmp_image_meta_dir / regular_inf_fname,
+            new_dirs=self._ota_tmp_image_meta_dir / dir_inf_fname,
+        )
```
Comment (obi-t4) [inline otaclient/app/create_standby/rebuild_mode.py:85]: Sorry, why is this modification necessary for this PR?

**Review by obi-t4 — COMMENTED** [review]
I left comment. I'd like to hear your comment.

File: otaclient/app/create_standby/rebuild_mode.py (hunk 2 above)
Comment (Bodong-Yang) [reply otaclient/app/create_standby/rebuild_mode.py:85]: **time-based retry strategy** depends on managing all the downloading tasks within the same **downloading tasks group**, to support **downloading tasks group**, delta needs to be pre-calculated before otaclient starts downloading. 
Delta generation is done by the create_standby module, so this API method is exposed by create_standby module.

Comment (Bodong-Yang) [reply otaclient/app/create_standby/rebuild_mode.py:85]: Please check `create_standby/interface.py` for more details.

File: otaclient/app/create_standby/rebuild_mode.py [hunk 3]
```diff
@@ -179,89 +117,45 @@ def _process_persistents(self):
...
-                "all process_regulars tasks are dispatched, wait for finishing..."
+            _mapper = RetryTaskMap(
+                self._process_regular,
+                self.delta_bundle.new_delta.items(),
+                title="process_regulars",
+                max_concurrent=cfg.MAX_CONCURRENT_PROCESS_FILE_TASKS,
+                backoff_max=cfg.CREATE_STANDBY_BACKOFF_MAX,
+                backoff_factor=cfg.CREATE_STANDBY_BACKOFF_FACTOR,
+                max_retry=cfg.CREATE_STANDBY_RETRY_MAX,
+                executor=pool,
```
Comment (obi-t4) [inline otaclient/app/create_standby/rebuild_mode.py:141]: It seems that `_process_regular` doesn't download anything, is this understanding correct?
If so, why `_process_regular` is wrapped with RetryTaskMap?

Comment (Bodong-Yang) [reply otaclient/app/create_standby/rebuild_mode.py:141]: `RetryTaskMap` can manage a group of tasks, not only manage downloading. It is a generic tasks manager actually. So in here, the `RetryTaskMap` is managing the processing regular file tasks.

File: otaclient/app/create_standby/common.py [hunk 4]
```diff
@@ -303,85 +298,164 @@ def _cal_and_prepare_old_slot_delta(self):
...
+            for _entry in map(RegularInf.parse_reginf, f):
+                self.total_regulars_num += 1
+                self._new.add_entry(_entry)
+                self._new_hash_list.add(_entry.sha256hash)
+        self._stats_collector.store.total_regular_files = self.total_regulars_num
+        # pre-load delta src's reginf if presented
+        if Path(delta_src_reg).is_file():
+            with open(delta_src_reg, "r") as f:
+                for _entry in map(RegularInf.parse_reginf, f):
+                    self._delta_src_path_hash_dict[str(_entry.path)] = _entry.sha256hash
```
Comment (obi-t4) [inline otaclient/app/create_standby/common.py:440]: I think this is the new feature. It should be the another PR. Is it well tested?
This feature changes the ota client behavior if `/opt/ota/image-meta/regulars.txt` exists or not.

File: otaclient/app/create_standby/rebuild_mode.py [hunk 5]
```diff
@@ -61,95 +42,52 @@
...
-            # TODO: define specific error for delta generator
-            raise UpdateDeltaGenerationFailed from e
+        delta_calculator = DeltaGenerator(
+            delta_src=self.active_slot_mp,
+            local_copy_dir=self._ota_tmp,
+            stats_collector=self.stats_collector,
+        )
+        delta_bundle = delta_calculator.calculate_and_process_delta(
+            delta_src_reg=Path(cfg.META_FOLDER) / regular_inf_fname,
+            new_reg=self._ota_tmp_image_meta_dir / regular_inf_fname,
```
Comment (obi-t4) [inline otaclient/app/create_standby/rebuild_mode.py:83]: In most of the case this is not a problem but it seems that `regular_inf_fname` is used for the **previous** metadata.jwt set. previous regular_inf_fname and current regular_inf_fname might be different.

**Review by obi-t4 — COMMENTED** [review]
I think `_delta_src_path_hash_dict` feature is not related retry strategy, so I'll disable this feature at least from this PR. What do you think. Any comment appreciate!

File: otaclient/app/create_standby/common.py [hunk 6]
```diff
@@ -221,63 +237,45 @@ class DeltaGenerator:
...
-        self.total_regulars_num = 0
-        with open(reginf_file, "r") as f, ProcessPoolExecutor() as pool:
-            for entry in pool.map(RegularInf.parse_reginf, f, chunksize=2048):
-                self.total_regulars_num += 1
-                self._new_delta.add_entry(entry)
-                self._hash_set.add(entry.sha256hash)
-
-        self._stats_collector.store.total_regular_files = self.total_regulars_num
+        self._stats_collector = stats_collector
+        self._delta_src_mp = delta_src
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/create_standby/common.py:255]: Ah, it means mount point. I can use the full name instead of abbreviation if needed.

File: otaclient/app/create_standby/common.py (hunk 4 above)
Comment (Bodong-Yang) [reply otaclient/app/create_standby/common.py:440]: @obi-t4  Yes, it can speed up the delta generating process if the current running slot has OTA image meta files saved in `/opt/ota/image-meta`. 
That is actually the leftover TODO for `DeltaGenerator` implementation. If needed I can remove it and re-add it in another PR.

File: otaclient/app/create_standby/rebuild_mode.py (hunk 5 above)
Comment (Bodong-Yang) [reply otaclient/app/create_standby/rebuild_mode.py:83]: @obi-t4 That is definitely needed to be concerned! I will strip away this part of implementation in `DeltaGenerator`, and re-introduce the new implementation in the future PR.

**Comment by Bodong-Yang** [comment]
> I think `_delta_src_path_hash_dict` feature is not related retry strategy, so I'll disable this feature at least from this PR. What do you think. Any comment appreciate!

I agree with you. I will strip away this part of implementation of `DeltaGenerator` for now and re-introduce it in the future PR.

File: otaclient/app/create_standby/common.py (hunk 4 above)
Comment (obi-t4) [reply otaclient/app/create_standby/common.py:440]: We need to do two test cases:
1. /opt/ota/image-meta/regulars.txt exists
2. /opt/ota/image-meta/regulars.txt doesn't exist
In this PR, I'd like to focus on the new retry-strategy.

**Comment by Bodong-Yang** [comment]
Long run testing result:
```
[2023-01-26 15:24:18,747][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=5974, self._last_failed_count=0, _total_retry_count=0
[2023-01-26 15:27:46,291][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=2026, self._last_failed_count=5974, _total_retry_count=1
[2023-01-26 15:28:36,791][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=442, self._last_failed_count=2026, _total_retry_count=2
[2023-01-26 15:28:58,593][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=150, self._last_failed_count=442, _total_retry_count=3
[2023-01-26 15:29:12,301][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=141, self._last_failed_count=150, _total_retry_count=4
[2023-01-26 15:29:27,236][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=141, self._last_failed_count=141, _total_retry_count=5
[2023-01-26 15:29:37,570][ERROR]-otaclient.app.common:_task_dispatcher:493,downloading_ota_files failed to finished, len(self._failed_tasks)=105, self._last_failed_count=141, _total_retry_count=6
```

And finally I resume the Internet connection and downloading finished as expected.

**Comment by Bodong-Yang** [comment]
@obi-t4 Finally the long running test with the latest changes is finished without problem, the PR is ready for review now.

File: otaclient/app/common.py [hunk 7]
```diff
@@ -306,83 +334,264 @@ def urljoin_ensure_base(base: str, url: str):
...
+                self = None  # resolve cycle reference
+
+    # API
+
+    def shutdown(self) -> bool:
+        """Set the internal status to SHUTDOWNED to trigger shutting down."""
+        with self._status_lock:
+            if self._status is not _RetryTaskMapStatus.RUNNING:
+                return False
+        self._status = _RetryTaskMapStatus.SHUTDOWNED
```
Comment (obi-t4) [inline otaclient/app/common.py:566]: this assignment should be in the `with` block.

File: otaclient/app/common.py [hunk 8]
```diff
@@ -306,83 +334,264 @@ def urljoin_ensure_base(base: str, url: str):
...
+                self._yield_que.put_nowait(_fut)
+            except Exception:  # ignored as not caused by task itself
+                pass
+
+            # if the only element in self._futs is _ROUNT_START_SENTINEL,
+            # it means that this retry round is finished.
+            if (
+                self._dispatcher_done_this_round.is_set()
+                and len(self._futs) == 1
+                and _ROUND_START_SENTINEL in self._futs
```
Comment (obi-t4) [inline otaclient/app/common.py:472]: I don't know why this `_ROUND_START_SENTINEL` is added in the self._futs set and checked `if len(self._futs) == 1`.
Why remove `self._futs.add(_ROUND_START_SENTINEL)`, and `len(self._futs) == 0` is not enough?

**Review by obi-t4 — COMMENTED** [review]
I left some comment, please have a check.
review is in progress.

**Review by obi-t4 — COMMENTED** [review]
I left come comment and review has been finished. Basically I think OK.

File: otaclient/app/common.py (hunk 8 above)
Comment (Bodong-Yang) [reply otaclient/app/common.py:472]: Actually this is a trick to let the collector generator awares of each retry round and prevent the collector from pre-maturely ending the collecting for this retry Round.
But yes the naming of this sentinel is kind of misleading, I will change it to `_RETRY_ROUND_SENTINEL`.

> len(self._futs) == 0 is not enough?

If the tasks finished faster than tasks being dispatched, there is a chance that the `self._futs` will be empty and new tasks are not yet dispatched and scheduled, causing the collector thinks it finishes its job in this retry round. I actually encountered the above situation when doing test, so I introduce this `_RETRY_ROUND_SENTINEL` mechanism to avoid pre-maturely round ending.

Comment (obi-t4) [reply otaclient/app/common.py:472]: > If the tasks finished faster than tasks being dispatched, there is a chance that the self._futs will be empty and new tasks are not yet dispatched and scheduled, causing the collector thinks it finishes its job in this retry round.

I think even if _ROUND_START_SENTINEL is introduced, the following condition will be satisfied on the above circumstance?
It means introducing `_ROUND_START_SENTINEL` can't solve this `finished faster than tasks being dispatched` condition.
```
and len(self._futs) == 1
and _ROUND_START_SENTINEL in self._futs
```

File: otaclient/app/create_standby/common.py [hunk 9]
```diff
@@ -303,85 +338,141 @@ def _cal_and_prepare_old_slot_delta(self):
...
+    def calculate_and_process_delta(
+        self,
+        *,
+        delta_src_reg: Path,  # path to delta_src regulars.txt, currently not used
+        new_reg: Path,  # path to new image regulars.txt
+        new_dirs: Path,  # path to dirs.txt
+    ) -> DeltaBundle:
+        with open(new_dirs, "r") as f:
+            for _dir in map(DirectoryInf, f):
+                self._new_dirs[_dir] = None
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/create_standby/common.py:450]: Emmmm for better code readability, I still think the `_new_dirs` should be bound to the inst, as `_new_dirs` is a part of the delta. 
Also, it is also a little strange for `_process_delta_src` to take only `_new_dirs` as input, but not take other inst variable as input. Actually `_process_delta_src` needs many inst variables to function, including `_new_dirs`, these used variables should have the same scope.

File: otaclient/app/common.py [hunk 10]
```diff
@@ -306,83 +334,264 @@ def urljoin_ensure_base(base: str, url: str):
...
+                if self._status is not _RetryTaskMapStatus.RUNNING:
+                    return
+                self._futs.discard(_fut)
+                self._task_collector_gen.send(_fut)
+
+            # NOTE: signal before examine the shutdown flag to prevent dead lock
+            # NOTE: pre-add a sentinel to indicate the collector about dispatch round end
+            self._futs.add(_ROUND_START_SENTINEL)  # type: ignore
+            if self._status is not _RetryTaskMapStatus.RUNNING:
+                return
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/common.py]: Updated in [c7c5cea](https://github.com/tier4/ota-client/pull/193/commits/c7c5ceaa431685553e134302beac8b68947c918f).

File: otaclient/app/common.py (hunk 8 above)
Comment (Bodong-Yang) [reply otaclient/app/common.py]: sentinel here is not a `Future` inst created by the executor.sumbit function, so it will not be discard by done_callback, only the collector generator can remove it.

Comment (obi-t4) [reply otaclient/app/common.py]: `self._futs.add(_RETRY_ROUND_SENTINEL)` is done before any task submit.
So I think checking

```
and len(self._futs) == 1
and _ROUND_START_SENTINEL in self._futs
```

and 

```
and len(self._futs) == 0
```

are logically same if _RETRY_ROUND_SENTINEL is not set to self._futs.add.

Comment (Bodong-Yang) [reply otaclient/app/common.py]: @obi-t4 Emmm you are right, let me reconsider again.

Comment (obi-t4) [reply otaclient/app/common.py]: I think we need to wait until length of self._iter and self._futs are the same in the done callback.

Comment (Bodong-Yang) [reply otaclient/app/common.py]: @obi-t4 The `RetryTaskMap` is fixed in [67e43e8](https://github.com/tier4/ota-client/pull/193/commits/67e43e8a490671d82b2a53bab9ba63e45a1c84e0), please have a check, thank you!

**Comment by Bodong-Yang** [comment]
@obi-t4 I confirmed that the bug I found is not related with this PR, and it is fixed by #195 , so I will not address the bug in this PR. Please continue the review, thank you!

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:
